File: convex_hull_trick.cpp
#include <bits/stdc++.h>
using namespace std;

struct Line {
    long long a, b;
    Line(long long a = 0, long long b = 0) : a(a), b(b) {
    }
    long long get(long long x) const {
        return a * x + b;
    }
    /**
     * y = a1x + b1
     * y = a2x + b2
     * a1x + b1 = a2x + b2
     * x = (b2 - b1) / (a1 - a2)
    */
    double intersect(const Line &l) const {
        return (double) (l.b - this->b) / (double) (this->a - l.a);
    }
};

struct HullTrickOperation {
    int pos, size;
    Line line;

    HullTrickOperation(int pos, int size, Line line) : pos(pos), size(size), line(line) {}
};

class HullTrick {
    private:
        bool min_hull;
        bool reverse;
        int size;
        Line *lines;
        stack<HullTrickOperation> operations;

        static bool compare(const Line &l1, const Line &l2, const Line &l3) {
            return l1.intersect(l3) >= l1.intersect(l2);
        }

        static bool compare_reverse(const Line &l1, const Line &l2, const Line &l3) {
            return l1.intersect(l3) <= l1.intersect(l2);
        }

        static bool cmp(double x, double y) {
            return x > y;
        }

        static bool cmp_reverse(double x, double y) {
            return x < y;
        }
    public:
        HullTrick(int N = 100000, bool min_hull = true, bool reverse = false) {
            this->min_hull = min_hull;
            this->reverse = reverse;
            this->lines = new Line[N];
            this->size = 0;
        }

        bool add(const Line &line, bool (*compare)(const Line &, const Line &, const Line &) = nullptr) {
            if(size < 2) {
                operations.push(HullTrickOperation(size, size, lines[size]));
                this->lines[size++] = line;
                return true;
            }

            if(compare == nullptr) {
                if(this->min_hull) {
                    compare = this->reverse ? compare_reverse : HullTrick::compare;
                } else {
                    compare = this->reverse ? HullTrick::compare : HullTrick::compare_reverse;
                }
            }

            int l = 1, r = size - 1;
            int new_pos = size;
            while(l <= r) {
                int m = (l + r) >> 1;
                if(compare(lines[m - 1], lines[m], line)) {
                    new_pos = m;
                    r = m - 1;
                } else {
                    l = m + 1;
                }
            }

            operations.push(HullTrickOperation(new_pos, size, lines[new_pos]));
            size = new_pos + 1;
            lines[new_pos] = line;
            return true;
        }

        void rollback() {
            if(operations.empty()) {
                return;
            }

            HullTrickOperation operation = operations.top();
            operations.pop();
            size = operation.size;
            lines[operation.pos] = operation.line;
        }

        long long query(long long target, bool (*cmp)(double, double) = nullptr) {
            if(cmp == nullptr) {
                if(this->min_hull) {
                    cmp = this->reverse ? HullTrick::cmp : HullTrick::cmp_reverse;
                } else {
                    cmp = this->reverse ? HullTrick::cmp_reverse : HullTrick::cmp;
                }
            }

            int l = 0, r = size - 1;
            long long ans = lines[l].get(target);

            while(l < r) {
                int m = (l + r) >> 1;
                long long x = lines[m].get(target);
                long long y = lines[m + 1].get(target);
                // cout << l << " " << r << " " << m << " " << x << " " << y << endl;
                if(cmp(x, y)) {
                    l = m + 1;
                } else {
                    r = m;
                }

                if(this->min_hull) {
                    ans = min(ans, min(x, y));
                } else {
                    ans = max(ans, max(x, y));
                }
            }

            return ans;
        }

        int getSize() {
            return size;
        }

        const Line &get(int i) {
            return lines[i];
        }

        void debug() {
            for(int i = 0; i < size; i++) {
                cout << "Line: " << lines[i].a << " " << lines[i].b << endl;
            }

            cout << endl;

            // while(!operations.empty()) {
            //     HullTrickOperation operation = operations.top();
            //     operations.pop();
            //     cout << "Operation: " << operation.pos << " " << operation.size << " " << operation.line.a << " " << operation.line.b << endl;
            // }
        }
};

int main() {
    Line lines[] = {Line(1, 0), Line(2, 3), Line(3, 1), Line(4, 2), Line(5, 0)};

    sort(lines, lines + 5, [](const Line &l1, const Line &l2) {
        return l1.a > l2.a;
    });

    HullTrick hull(5, true, true);
    for(int i = 0; i < 5; i++) {
        hull.add(lines[i]);
        // cout << "Added" << endl;
    }

    cout << "Added" << endl;

    cout << "Expect 0: " << hull.query(1) << endl;
    cout << hull.query(1) << endl;

    hull.debug();
}

File: convex_hull_trick.py
import matplotlib.pyplot as plt

class Line:
    def __init__(self, a = 0, b = 0):
        self.a = a
        self.b = b

    def eval(self, x):
        return self.a * x + self.b

    def __str__(self):
        return f"f(x) = {self.a}x + {self.b}"

    def __repr__(self):
        return str(self)
    
    def intersection(self, other: "Line"):
        x = (other.b - self.b) / (self.a - other.a)
        return x, self.f(x)
    
class HullTrick:
    def __init__(self, N = 100000, is_min_hull = True, reverse = True) -> None:
        self.is_min_hull = is_min_hull
        self.reverse = reverse
        self._N = N
        self.size = 0
        self.lines = [Line(0, 0) for _ in range(N)]
        self.operations = []
        
    def __add_compare(l1: Line, l2: Line, l3: Line) -> bool:
        return l1.intersection(l3)[0] >= l1.intersection(l2)[0]
    def __add_compare_reverse(l1: Line, l2: Line, l3: Line) -> bool:
        return l1.intersection(l3)[0] <= l1.intersection(l2)[0]
    
    def __get_add_comparer(self):
        if self.is_min_hull:
            return self.__add_compare if not self.reverse else self.__add_compare_reverse
        
        return self.__add_compare if self.reverse else self.__add_compare_reverse 
    
    def comparer(x, y):
        return x < y
    
    def comparer_reverse(x, y):
        return x > y
    
    def __get_comparer(self):
        if self.is_min_hull:
            return self.comparer if not self.reverse else self.comparer_reverse
        return self.comparer if self.reverse else self.comparer_reverse
        
    def add(self, line: Line):
        if self.size < 2:
            self.lines[self.size] = line
            self.operations.append((self.size, self.size, self.lines[self.size]))
            self.size += 1
            return

        comparer = self.__get_add_comparer()
        L = 1
        R = self.size - 1
        POS = self.size
        while L < R:
            M = (L + R) // 2
            if comparer(self.lines[M - 1], self.lines[M], line):
                POS = M
                L = M + 1
            else:
                R = M
        
        self.operations.append((POS, self.size, self.lines[POS]))
        self.size = POS + 1
        self.lines[POS] = line
        
    def rollback(self):
        if len(self.operations) == 0:
            return
        op = self.operations[-1]
        self.size = op[0]
        self.lines[op[0]] = op[2]
        self.operations.pop()

    def get(self, x):
        comparer = self.__get_comparer()
        L = 0
        R = self.size - 1
        res = self.lines[L].eval(x)
        while L < R:
            M = (L + R) // 2
            e1 = self.lines[M].eval(x)
            e2 = self.lines[M + 1].eval(x)
            if comparer(e1, e2):
                L = M + 1
            else:
                R = M
                
            res = min(res, e1, e2) if self.is_min_hull else max(res, e1, e2)
    
def plot_lines(lines: list[Line]):
    x = [-500, 500]
    plt.ylim(-500, 500)
    for line in lines:
        y = [line.f(x[0]), line.f(x[1])]
        plt.plot(x, y, label=str(line))
    plt.legend()
    plt.show()

lines = [Line(0, 0), Line(-21, 542), Line(-20, 206)]
plot_lines(lines)

stack: list[Line] = []
lines.sort(key=lambda line: line.a, reverse=True)
print(lines)

for line in lines:
    if len(stack) < 2:
        stack.append(line)
        continue
    while len(stack) >= 2:
        l1 = stack[-2]
        l2 = stack[-1]
        l3 = line
        
        if l1.intersection(l3)[0] > l1.intersection(l2)[0]:
            break
        stack.pop()
    
    stack.append(line)
    
plot_lines(stack)
        